Source/SwiftLintBuiltInRules/Rules/Idiomatic/FileNameRule.swift

```diff
@@ -56,33 +58,118 @@ struct FileNameRule: OptInRule, SourceKitFreeRule {
+    private func visitNode(_ node: some TypeNameCollectible) -> SyntaxVisitorContinueKind {
+        let name = node.typeName
+        let fullyQualifiedName = (ancestorNames + [name]).joined(separator: ".")
+        names.insert(fullyQualifiedName)
+
+        if !requireFullyQualifiedNames {
+            names.insert(name)
+        }
+
+        ancestorNames.push(node.typeName)
+        return .visitChildren
+    }
+
+    /// Removes a node's type name as an ancestor type name once all of its children have been visited
+    private func visitNodePost(_: some TypeNameCollectible) {
+        ancestorNames.pop()
+    }
+
+    override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
+    }
+
     override func visitPost(_ node: ClassDeclSyntax) {
-        names.insert(node.name.text)
+        visitNodePost(node)
+    }
+
+    override func visit(_ node: ActorDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
     }
 
     override func visitPost(_ node: ActorDeclSyntax) {
-        names.insert(node.name.text)
+        visitNodePost(node)
+    }
+
+    override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
     }
 
     override func visitPost(_ node: StructDeclSyntax) {
-        names.insert(node.name.text)
+        visitNodePost(node)
+    }
+
+    override func visit(_ node: TypeAliasDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
     }
 
     override func visitPost(_ node: TypeAliasDeclSyntax) {
-        names.insert(node.name.text)
+        visitNodePost(node)
+    }
+
+    override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
     }
 
     override func visitPost(_ node: EnumDeclSyntax) {
-        names.insert(node.name.text)
+        visitNodePost(node)
+    }
+
+    override func visit(_ node: ProtocolDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
     }
 
     override func visitPost(_ node: ProtocolDeclSyntax) {
-        names.insert(node.name.text)
+        visitNodePost(node)
+    }
+
+    override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
+        visitNode(node)
     }
 
     override func visitPost(_ node: ExtensionDeclSyntax) {
-        names.insert(node.extendedType.trimmedDescription)
+        visitNodePost(node)
+    }
+}
+
+/// A protocol for types that have a type name that can be collected
+private protocol TypeNameCollectible {
+    var typeName: String { get }
+}
+
+extension TypeNameCollectible where Self: NamedDeclSyntax {
+    var typeName: String {
+        name.trimmedDescription
+    }
+}
+extension ClassDeclSyntax: TypeNameCollectible {}
+extension ActorDeclSyntax: TypeNameCollectible {}
+extension StructDeclSyntax: TypeNameCollectible {}
+extension TypeAliasDeclSyntax: TypeNameCollectible {}
+extension EnumDeclSyntax: TypeNameCollectible {}
+extension ProtocolDeclSyntax: TypeNameCollectible {}
```

Comment:
That's a lot of overhead only to spare one more method. Honestly, I'd prefer to go with a `visit(name:String)` and a `visit(node:NamedDeclSyntax)`. The latter can just redirect to the first after extracting the name. And `visit(_: ExtensionDeclSyntax)` can call it directly.

Reply:
No problem. I'll go this route instead